docs/proposals/improve-addons.md

```diff
@@ -0,0 +1,350 @@
+- [Implementation](#implementation)
+  - [Architecture](#architecture)
+  - [Components](#components)
+    - [AddonCatalogueDistributor](#addoncataloguedistributor)
+    - [AddonCatalogue CR](#addoncatalogue-cr)
+    - [AddonController](#addoncontroller)
+    - [AddonInstallation CR](#addoninstallation-cr)
+    - [CoreAddonController](#coreaddoncontroller)
+    - [AddonInstallationController](#addoninstallationcontroller)
+  - [Common User Flows](#common-user-flows)
+  - [Alternatives considered](#alternatives-considered)
+- [Glossary](#glossary)
+  - [CoreAddon (old name: Default **Addon**)](#coreaddon-old-name-default-addon)
+  - [OptionalAddon (old name: Accessible Addon)](#optionaladdon-old-name-accessible-addon)
+  - [CustomerAddon (old name: Custom Addon)](#customeraddon-old-name-custom-addon)
+- [Task & effort](#task--effort)
+
+## Introduction
+
+This proposal introduces improvements to the current [Addon system](https://docs.kubermatic.com/kubermatic/v2.18/guides/addons/). It aims to make current addons more flexible and user-friendly. Additionally, it simplifies the usage of addons we ship with (almost) every cluster, so-called "Default Addons".
+
+## Motivation and Background
+
+Currently, KKP only supports one mechanism for addons: For this, all addons are stored in a single docker image that is configured in KKP. Afterwards, an initContainer is being run that starts the image and copies all addons onto the seed-controller-managers local file-system. From there, addon manifests are being rendered and rolled out to user clusters.
+
+The current implementation of Addons has some flaws that make the usage and maintenance of Addons cumbersome for cluster administrators:
+
+- To add “Custom Addons” the cluster-admin needs to bundle manifests and templates for these Addons in a Docker image that is based on the official addon-manifest docker image by Kubermatic.
+- It’s currently only possible to deploy multiple instances of the same Addon when using a workaround
+- Addons are strictly tied to KKP versions due to the composition logic of the manifest image tag; for any provided manifest image the tag gets overwritten with `:<KKP VERSION>\[-CUSTOM SUFFIX\]`
+- Mandatory core components of user clusters are named “Default Addons” which implies that these can be deleted by Cluster Admins
+- Complex Addons can leave a trail of reconciliation errors until successfully deployed because all manifest files are applied at the same time
+- Creating a cluster is a two-step process; firstly you have to create a cluster and only  afterward you can select custom Addons
+
+*note: PS team report that some clients are using weekly KKP update in dev environments, consequently they need to rebuild the addon image every week which is painful and error-prone.*
+
+## Goals
+
+- Decouple custom addons from the KKP lifecycle
+- Handle dependency **inside** an Addon (e.g. CRD need to be created before CRs, namespace needs to be created before other objects)
```

Comment:
How is this goal achieved in the proposal? Maybe I missed it, but I couldn't find it. The only thing I found is `RequiredResourceTypes`, and my understanding is that this is actually defining outside dependencies (prerequisites for successfully installing the addon).

Reply:
Ah interesting. Maybe I misunderstood our current implementation. I thought our current implementation would do exactly that: ensure that CRDs and namespaces are deployed before any other manifests.
cc @vgramer: This is something I think we have to look in the code of the current implementation. But yes, if the current system is not able to install CRDs before deploying resources, then we should include that capability in our proposal.

On the other side: For the helm rendering method, handling dependencies is done directly via helm; So theoretically this should come out of the box (of course during implementation we will have to see how well it works in practice)

Reply:
Ah, I was more thinking of a flexible dependency management system within the addon, like https://carvel.dev/kapp/docs/latest/apply-ordering/. That's probably fairly complex to pull off, but I misunderstood the goal then :)

Reply:
@SimonTheLeg yes the `RequiredResourceTypes` works as @embik explain. it's a pre-requisite to render that addon's manifests.

Concerning the `Handle dependency **inside** an Addon ` I think too that `helm` should do the job